### PR TITLE
accounts refactoring

### PR DESCRIPTION
**Pull request by pchrysochoidis**
## Description 

Enables storing public data (address, public key, etc) for accounts to allow showing account data even when the accounts are locked. (this only runs through the test `#/accounts-dev` page for dev mode and when `NEW_ACCOUNTS_ENABLED` constant is set to true)

* adds new Account classes that handle specific account operations based on the account type
* also adds AccountSources that mainly used to store shared data between same type of accounts. eg. MnemonicAccountSource that stores the mnemonic and can be shared across the mnemonic accounts
* accounts now can be read unlocked (unable to do write operations) but able to show balances etc. This will be mainly a UI privacy feature where we check the `lastUnlockedOn` to be less than max allowed
* store new accounts using indexedDB with dexie js
    * to address any concerns of browser clearing the indexedDB data
      * added `unlimitedStorage` permission
      * and a mechanism to import and export the db from local storage. After adding a new account or account source indexedDB is being saved to `local storage` and when the database opens we check to import the exported data
* integrate Sentry in the service worker context, to allow logging errors about indexedDB being cleared from the browser/user
* added some test pages to test adding/removing accounts that will be deleted later
* each account and account source unlocked individually and stores any sensitive data in session storage encrypted
* currently accounts and account sources get locked only manually (or when browser clears the session storage) - will revisit this later when we have a better idea about locking accounts

https://github.com/MystenLabs/sui/assets/10210143/f0a07233-563b-4b41-b6f8-ebab3d17abca

https://github.com/MystenLabs/sui/assets/10210143/eed04ad2-2a23-4ca7-abf6-50a0b9880864

https://github.com/MystenLabs/sui/assets/10210143/a7a776ab-d66d-4f84-b418-b1925c15b290

closes APPS-1099

## Test Plan 

change `NEW_ACCOUNTS_ENABLED` to true and visit `#/accounts-dev` page

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: apps/wallet/src/background/storage-utils.ts [hunk 1]
```diff
@@ -1,17 +1,31 @@
...
+export type LocalStorageKey = 'active-account-id-key';
+
 const SESSION_STORAGE: Storage.LocalStorageArea | null =
 	// @ts-expect-error chrome
 	global?.chrome?.storage?.session || null;
 
 async function getFromStorage<T>(
 	storage: Storage.LocalStorageArea,
-	key: string,
+	key: LocalStorageKey | string, // TODO: try to log all LocalStorage keys and remove string
```
Comment (pchrysochoidis) [inline apps/wallet/src/background/storage-utils.ts:28]: Remove this

File: apps/wallet/src/background/db.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		this.version(1).stores({
+			accountSources: 'id, type',
+			accounts: 'id, type, address, sourceID',
+			settings: 'setting',
+		});
+	}
+}
+
+async function init() {
+	const db = new DB();
```
Comment (pchrysochoidis) [inline apps/wallet/src/background/db.ts:30]: I updated the way we import the DB from a backup to account for different schema versions. Now we can import a backup the has a previous version that the current, and in case there are any upgrades they will run as expected.

**Review by williamrobertson13 — APPROVED** [review]
I took a few passes at this, and it overall looks good to me and seems like a pretty reasonable way to go about handling accounts given our current setup

(sorry for the delay on the review btw!)

File: apps/wallet/src/background/account-sources/AccountSource.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	protected setEphemeralValue(value: V) {
+		return setEphemeralValue(this.id, value);
+	}
+
+	protected clearEphemeralValue() {
+		return clearEphemeralValue(this.id);
+	}
+}
+
+export interface AccountSourceSerialized {
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/account-sources/AccountSource.ts:51]: Yeah probably not - we might need to hide fields from UI

File: apps/wallet/src/ui/app/index.tsx [hunk 4]
```diff
@@ -72,6 +74,27 @@ const App = () => {
 
 	return (
 		<Routes>
+			{/* this is used only for making dev work on refactoring accounts easier - TODO: remove when work is done ----> */}
+			{process.env.NODE_ENV === 'development' && NEW_ACCOUNTS_ENABLED ? (
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/ui/app/index.tsx:78]: Yeah it was done before we decided that. But I will remove it later to avoid breaking the wallet now. It will be easy to remove anyway

File: apps/wallet/src/shared/cryptography/keystore.ts [hunk 5]
```diff
@@ -5,8 +5,20 @@ import {
...
 	decrypt as metamaskDecrypt,
 } from '@metamask/browser-passworder';
+import { randomBytes } from '@noble/hashes/utils';
 
-type Serializable = string | number | boolean | { [index: string]: Serializable } | Serializable[];
+// we use this password + a random one for each time we store the encrypted
+// vault to session storage
+const PASSWORD =
+	process.env.WALLET_KEYRING_PASSWORD ||
+	'344c6f7d04a65c24f35f5c710b0e91e2f2e2f88c038562622d5602019b937bc2c2aa2821e65cc94775fe5acf2fee240d38f1abbbe00b0e6682646a4ce10e908e';
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/shared/cryptography/keystore.ts:14]: You mean the constant? or in general how we use the `PASSWORD`?

the constant is there just to make sure we have a password in case process.env.WALLET_KEYRING_PASSWORD is empty

and we use this to encrypt values in session storage with a known password which changes between releases (to make it a bit more difficult to decrypt the values if someone gets access to them)

File: apps/wallet/src/ui/app/ApiProvider.ts [hunk 6]
```diff
@@ -90,44 +91,46 @@ export default class ApiProvider {
...
-				// behavior due to the reactive nature of managing Ledger connections
-				// and displaying relevant UI updates. Refactoring ApiProvider to
-				// not be a global instance would help out here, but that is also
-				// a non-trivial task because we need access to ApiProvider in the
-				// background script as well.
-				throw new Error("Signing with Ledger via ApiProvider isn't supported");
-			default:
-				throw new Error('Encountered unknown account type');
+		if (
+			[AccountType.DERIVED, AccountType.IMPORTED, 'mnemonic-derived', 'imported'].includes(
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/ui/app/ApiProvider.ts:101]: I added it, but ts will force as to do it when we remove the current Account classes anyway

File: apps/wallet/src/manifest/manifest.json [hunk 7]
```diff
@@ -6,7 +6,7 @@
 	"background": {
 		"service_worker": "background.js"
 	},
-	"permissions": ["storage", "tabs", "alarms"],
+	"permissions": ["storage", "tabs", "alarms", "unlimitedStorage"],
```
Comment (mamos-mysten) [inline apps/wallet/src/manifest/manifest.json:9]: we might want to include something in our release notes to inform users why this was added

File: apps/wallet/src/background/account-sources/QredoAccountSource.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+		const { apiUrl } = await this.getStoredData();
+		return (
+			await new QredoAPI(this.id, apiUrl).createAccessToken({
+				refreshToken,
+			})
+		).access_token;
+	}
+
+	get apiUrl() {
+		return this.getStoredData().then(({ apiUrl }) => apiUrl);
```
Comment (mamos-mysten) [inline apps/wallet/src/background/account-sources/QredoAccountSource.ts:148]: we don't use `cachedData` here is that intended?

**Review by mamos-mysten — APPROVED** [review]
Took a first pass at reviewing this, there's a lot here :) sorry for the delay in reviewing 

A couple general notes / questions: 

-  We have comments in some places, but I think adding more comments especially on interfaces and abstract class properties (`Account` , `AccountSource`) would really help readability
- Adding tests would help enforce boundaries as well as make clear the expected functionality of each component
- Do we have any fallback strategies if we can't read / write from IndexedDB? Should we? 
- Still not 100% clear on the intended difference between `AccountSource` and `Account` classes

File: apps/wallet/src/background/account-sources/AccountSource.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	abstract toUISerialized(): Promise<AccountSourceSerializedUI>;
+	abstract isLocked(): Promise<boolean>;
+	abstract lock(): Promise<void>;
+
+	protected async getStoredData() {
+		const data = await (await getDB()).accountSources.get(this.id);
+		if (!data) {
+			throw new Error(`Account data not found. (id: ${this.id})`);
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/account-sources/AccountSource.ts:33]: Probably it will never happen, but if so it will surface to the user

File: apps/wallet/src/background/account-sources/MnemonicAccountSource.ts [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+import { makeUniqueKey } from '../storage-utils';
+import {
+	getRandomEntropy,
+	entropyToSerialized,
+	entropyToMnemonic,
+	validateEntropy,
+} from '_shared/utils/bip39';
+import { decrypt, encrypt } from '_src/shared/cryptography/keystore';
+
+type DataDecryptedV0 = {
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/account-sources/MnemonicAccountSource.ts:28]: Yeah not necessary, I was thinking about any future updates - I will rename it

File: apps/wallet/src/background/account-sources/MnemonicAccountSource.ts [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+		}
+		let index = 0;
+		let derivationPath = '';
+		let temp;
+		do {
+			temp = this.makeDerivationPath(index++);
+			if (!derivationPathMap[temp]) {
+				derivationPath = temp;
+			}
+		} while (derivationPath === '' && index < 10000);
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/account-sources/MnemonicAccountSource.ts:166]: I added this for some kind of safety in case something goes wrong we won't get in an infinite loop. And chose 10000 as big enough that wont limit users from creating accounts and not very big to be noticeable for any error/infinite loop case

File: apps/wallet/src/background/account-sources/QredoAccountSource.ts [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+					{
+						apiUrl: await anAccountSource.apiUrl,
+						organization: await anAccountSource.organization,
+						origin: await anAccountSource.origin,
+						service: await anAccountSource.service,
+					},
+					{ id: dataSerialized.id, apiUrl, organization, origin, service },
+				)
+			) {
+				throw new Error('Qredo account source already exists');
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/account-sources/QredoAccountSource.ts:82]: Based on the way this is currently used throwing here I think is better. To make clear we can't store the new qredo source because it's a duplicate. There is check before calling this [here](https://github.com/MystenLabs/sui/blob/186ff82ba7f9e9d2b265df29c96c10b98468bffc/apps/wallet/src/background/qredo/index.ts#L235) so the error will probably never throw.

File: apps/wallet/src/ui/app/pages/qredo-connect/components/SelectQredoAccountsSummaryCard.tsx [hunk 13]
```diff
@@ -1,14 +1,41 @@
...
 import Loading from '_components/loading';
+import { isQredoAccountSerializedUI } from '_src/background/accounts/QredoAccount';
+import { isSerializedQredoAccount } from '_src/background/keyring/Account';
+import { NEW_ACCOUNTS_ENABLED } from '_src/shared/constants';
 import { type Wallet } from '_src/shared/qredo-api';
+import { useAccounts as useAccountsV2 } from '_src/ui/app/hooks/accounts-v2/useAccounts';
+import { useAccounts } from '_src/ui/app/hooks/useAccounts';
 import { Link } from '_src/ui/app/shared/Link';
 
+function useQredoConnectedAccounts() {
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/ui/app/pages/qredo-connect/components/SelectQredoAccountsSummaryCard.tsx:18]: Yeah, I will move accounts-v2 hook to the existing ones in future prs since we are ok on breaking the wallet

File: apps/wallet/src/background/accounts/Account.ts [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+
+	protected setEphemeralValue(value: V) {
+		if (!value) {
+			return;
+		}
+		return setEphemeralValue(this.id, value);
+	}
+
+	protected clearEphemeralValue() {
+		return clearEphemeralValue(this.id);
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/accounts/Account.ts:91]: Yeah can be one or many so I just kept it singular, like one value that can be object with many fields

File: apps/wallet/src/background/accounts/index.ts [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+	if (LedgerAccount.isOfType(account)) {
+		return new LedgerAccount({ id: account.id, cachedData: account });
+	}
+	if (QredoAccount.isOfType(account)) {
+		return new QredoAccount({ id: account.id, cachedData: account });
+	}
+	throw new Error(`Unknown account of type ${account.type}`);
+}
+
+export async function getAllAccounts(filter?: { sourceID: string }) {
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/accounts/index.ts:38]: Yeah we could extend it in we need it. But IIRC I didn't need it for now

File: apps/wallet/src/ui/app/helpers/sentry.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+		getSentryConfig({
+			integrations: [new Sentry.BrowserTracing()],
+			tracesSampler: () => {
+				return growthbook.getFeatureValue('wallet-sentry-tracing', 0);
+			},
+		}),
+	);
+}
+
+// expand this breadcrumb
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/ui/app/helpers/sentry.ts:19]: Just removed this code. We don't use it anywhere

File: apps/wallet/src/background/account-sources/index.ts [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+	return toAccountSource(serializedAccountSource);
+}
+
+export async function getAllSerializedUIAccountSources() {
+	return Promise.all(
+		(await getAccountSources()).map((anAccountSource) => anAccountSource.toUISerialized()),
+	);
+}
+
+async function createAccountSource({
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/account-sources/index.ts:56]: I was thinking to keep it generic in case we add more sources. I will see how it goes and probably change it later

File: apps/wallet/src/background/account-sources/QredoAccountSource.ts [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+			}
+		}
+		await (await getDB()).accountSources.put(dataSerialized);
```
[the remark this answers is not in the document]
Comment (pchrysochoidis) [reply apps/wallet/src/background/account-sources/QredoAccountSource.ts:85]: Yeah that was needed in the storage migration work and I did it there. (in order to avoid conflicts I wont do it here as well)

File: apps/wallet/src/background/account-sources/QredoAccountSource.ts (hunk 8 above)
Comment (pchrysochoidis) [reply apps/wallet/src/background/account-sources/QredoAccountSource.ts:148]: Yeah cache data for accounts was done mainly as optimisation and here it didn't seem necessary. (we can do it later if necessary)

**Comment by pchrysochoidis** [comment]
> We have comments in some places, but I think adding more comments especially on interfaces and abstract class properties (Account , AccountSource) would really help readability

Yeah I will add some in future PRs

> Adding tests would help enforce boundaries as well as make clear the expected functionality of each component

You are right, I am thinking to add e2e tests as we build the views and with those will cover this functionality as well.

> Do we have any fallback strategies if we can't read / write from IndexedDB? Should we?

I tested it in incognito and worked fine, we might see a few rare cases but we can address any issues like this if the come up  

> Still not 100% clear on the intended difference between AccountSource and Account classes

Yeah AccountSources is very similar to Accounts, the main reason for them is to hold any common data between similar accounts eg Mnemonic between mnemonic accounts and avoid duplicating it across the accounts. Also they help some functionality specific to different types of accounts.
